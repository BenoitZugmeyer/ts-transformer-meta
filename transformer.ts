import * as ts from "typescript"
import * as path from "path"

const indexTs = path.join(__dirname, "index.ts")

export default function transformer(
  program: ts.Program,
): ts.TransformerFactory<ts.SourceFile> {
  return (context: ts.TransformationContext) => (file: ts.SourceFile) =>
    visitSourceFileAndChildren(file, program, context)
}

function visitSourceFileAndChildren(
  sourceFile: ts.SourceFile,
  program: ts.Program,
  context: ts.TransformationContext,
) {
  return ts.visitEachChild(
    sourceFile,
    childNode => visitNodeAndChildren(childNode, program, context),
    context,
  )
}

function visitNodeAndChildren(
  node: ts.Node,
  program: ts.Program,
  context: ts.TransformationContext,
): ts.Node {
  return ts.visitEachChild(
    transformNode(node, program),
    childNode => visitNodeAndChildren(childNode, program, context),
    context,
  )
}

function transformNode(node: ts.Node, program: ts.Program): ts.Node {
  const typeChecker = program.getTypeChecker()
  if (!isKeysCallExpression(node, typeChecker)) {
    return node
  }
  if (!node.typeArguments) {
    return ts.createArrayLiteral([])
  }
  const type = typeChecker.getTypeFromTypeNode(node.typeArguments[0])
  return generateTypeMeta(typeChecker, type)
}

function generateTypeMeta(
  typeChecker: ts.TypeChecker,
  type: ts.Type,
): ts.Expression {
  const props = [
    ts.createPropertyAssignment("flags", ts.createLiteral(type.flags)),
  ]
  if (type.isClassOrInterface()) {
    props.push(
      ts.createPropertyAssignment(
        "objectFlags",
        ts.createLiteral(type.objectFlags),
      ),
    )
    props.push(
      ts.createPropertyAssignment(
        "properties",
        ts.createArrayLiteral(
          type.getProperties().map(property => {
            return ts.createObjectLiteral(
              [
                ts.createPropertyAssignment(
                  "name",
                  ts.createLiteral(property.name),
                ),
                ts.createPropertyAssignment(
                  "value",
                  generateTypeMeta(
                    typeChecker,
                    typeChecker.getTypeAtLocation(property.valueDeclaration),
                  ),
                ),
                ts.createPropertyAssignment(
                  "optional",
                  ts.createLiteral(
                    Boolean(property.flags & ts.SymbolFlags.Optional),
                  ),
                ),
              ],
              true,
            )
          }),
        ),
      ),
    )
  }
  if (type.isUnionOrIntersection()) {
    props.push(
      ts.createPropertyAssignment(
        "types",
        ts.createArrayLiteral(
          type.types.map(type => generateTypeMeta(typeChecker, type)),
        ),
      ),
    )
  }
  if (type.isLiteral()) {
    props.push(
      ts.createPropertyAssignment("value", ts.createLiteral(type.value)),
    )
  }
  return ts.createObjectLiteral(props, true)
}

function isKeysCallExpression(
  node: ts.Node,
  typeChecker: ts.TypeChecker,
): node is ts.CallExpression {
  if (!ts.isCallExpression(node)) return false
  const signature = typeChecker.getResolvedSignature(node)
  if (!signature) return false
  const declaration = signature.getDeclaration()
  return Boolean(
    declaration &&
      declaration.getSourceFile().fileName === indexTs &&
      declaration.kind === ts.SyntaxKind.FunctionDeclaration &&
      declaration.name &&
      declaration.name.getText() === "meta",
  )
}
