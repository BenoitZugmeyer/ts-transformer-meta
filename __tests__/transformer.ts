import * as ts from "typescript"
import { existsSync, readFileSync } from "fs"
import * as path from "path"
import transformer from "../transformer"

class VirtualHost implements ts.CompilerHost {
  files = new Map<string, string>()
  outputFiles = new Map<string, string>()
  modulesResolution = new Map<string, string>()

  getSourceFile(
    fileName: string,
    languageVersion: ts.ScriptTarget,
    _onError?: (message: string) => void,
  ): ts.SourceFile | undefined {
    const sourceText = this.readFile(fileName)
    return sourceText === undefined
      ? undefined
      : ts.createSourceFile(fileName, sourceText, languageVersion)
  }
  getDefaultLibFileName() {
    return path.join(path.dirname(require.resolve("typescript")), "lib.d.ts")
  }
  writeFile(fileName: string, content: string) {
    this.outputFiles.set(fileName, content)
  }
  getCurrentDirectory() {
    return "/"
  }
  getDirectories(_path: string) {
    return []
  }
  getCanonicalFileName(fileName: string) {
    return fileName
  }
  getNewLine() {
    return "\n"
  }
  useCaseSensitiveFileNames() {
    return true
  }
  fileExists(fileName: string): boolean {
    if (this.files.has(fileName)) {
      return true
    }
    return existsSync(fileName)
  }
  readFile(fileName: string): string | undefined {
    if (this.files.has(fileName)) {
      return this.files.get(fileName)
    }
    try {
      return readFileSync(fileName, { encoding: "utf-8" })
    } catch (e) {}
  }
  resolveModuleNames(
    moduleNames: string[],
    _containingFile: string,
  ): ts.ResolvedModule[] {
    return moduleNames.map(name => {
      const resolved = this.modulesResolution.get(name)
      if (!resolved) {
        throw new Error(`Unable to resolve module ${name}`)
      }
      return {
        resolvedFileName: resolved,
      }
    })
  }
}

function throwIfDiagnostics(
  diagnostics: ReadonlyArray<ts.Diagnostic>,
  host: ts.FormatDiagnosticsHost,
) {
  if (diagnostics.length > 0) {
    const formatedDiagnostics = ts.formatDiagnostics(diagnostics, host)
    throw new Error(formatedDiagnostics)
  }
}

function compile(input: string): string | undefined {
  const options: ts.CompilerOptions = {
    module: ts.ModuleKind.ES2015,
    target: ts.ScriptTarget.ES2018,
  }
  const host = new VirtualHost()
  host.files.set("input.ts", input)
  host.modulesResolution.set(
    "ts-transformer-meta",
    path.join(__dirname, "..", "index.ts"),
  )
  host.modulesResolution.set("typescript", require.resolve("typescript"))
  const program = ts.createProgram(["input.ts"], options, host)

  throwIfDiagnostics(
    [
      ...program.getOptionsDiagnostics(),
      ...program.getGlobalDiagnostics(),
      ...program.getSyntacticDiagnostics(),
      ...program.getSemanticDiagnostics(),
      ...program.getDeclarationDiagnostics(),
      ...program.getConfigFileParsingDiagnostics(),
    ],
    host,
  )

  const emitResult = program.emit(undefined, undefined, undefined, undefined, {
    before: [transformer(program)],
  })

  throwIfDiagnostics(emitResult.diagnostics, host)

  return host.outputFiles.get("input.js")
}

test("optional property", () => {
  expect(
    compile(`
    import { meta } from 'ts-transformer-meta'
    interface Foo {
      a?: number
    }
    meta<Foo>()
    `),
  ).toMatchSnapshot()
})

test("nested interfaces", () => {
  expect(
    compile(`
    import { meta } from 'ts-transformer-meta'
    interface Bar {
      b: number
    }
    interface Foo {
      a: Bar
    }
    meta<Foo>()
    `),
  ).toMatchSnapshot()
})

test("classes", () => {
  expect(
    compile(`
    import { meta } from 'ts-transformer-meta'
    class Biz {
      get foo() {
        return 42
      }
    }
    meta<Biz>()
    `),
  ).toMatchSnapshot()
})

test("union", () => {
  expect(
    compile(`
    import { meta } from 'ts-transformer-meta'
    meta<"foo" | "bar">()
    `),
  ).toMatchSnapshot()
})

test("enum", () => {
  expect(
    compile(`
    import { meta } from 'ts-transformer-meta'
    enum A {
      Foo,
      Bar,
    }
    meta<A>()
    `),
  ).toMatchSnapshot()
})
