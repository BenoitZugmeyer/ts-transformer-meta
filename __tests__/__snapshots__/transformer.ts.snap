// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classes 1`] = `
"import { meta } from 'ts-transformer-meta';
class Biz {
    get foo() {
        return 42;
    }
}
({
    flags: 524288,
    objectFlags: 5,
    properties: [{
            name: \\"foo\\",
            value: {
                flags: 8
            },
            optional: false
        }]
});
"
`;

exports[`enum 1`] = `
"import { meta } from 'ts-transformer-meta';
var A;
(function (A) {
    A[A[\\"Foo\\"] = 0] = \\"Foo\\";
    A[A[\\"Bar\\"] = 1] = \\"Bar\\";
})(A || (A = {}));
({
    flags: 1049600,
    types: [{
            flags: 1280,
            value: 0
        }, {
            flags: 1280,
            value: 1
        }]
});
"
`;

exports[`nested interfaces 1`] = `
"import { meta } from 'ts-transformer-meta';
({
    flags: 524288,
    objectFlags: 2,
    properties: [{
            name: \\"a\\",
            value: {
                flags: 524288,
                objectFlags: 2,
                properties: [{
                        name: \\"b\\",
                        value: {
                            flags: 8
                        },
                        optional: false
                    }]
            },
            optional: false
        }]
});
"
`;

exports[`optional property 1`] = `
"import { meta } from 'ts-transformer-meta';
({
    flags: 524288,
    objectFlags: 2,
    properties: [{
            name: \\"a\\",
            value: {
                flags: 8
            },
            optional: true
        }]
});
"
`;

exports[`union 1`] = `
"import { meta } from 'ts-transformer-meta';
({
    flags: 1048576,
    types: [{
            flags: 128,
            value: \\"foo\\"
        }, {
            flags: 128,
            value: \\"bar\\"
        }]
});
"
`;
